### DISCUSSION TITLE
what is the purpose of `ggml_half2 dm` in quantization structure?

### DISCUSSION THREAD

**Discussion by PenutChen** [ggml-org/ggml#887] [Q&A] [answered]
I notice that there is `ggml_half2 dm` in many k-quant structures, e.g. [block_q2_K](https://github.com/ggerganov/ggml/blob/a3c0188a4b5d3dec052ff87c9f773baa53631d70/src/ggml-common.h#L210-L220) and [block_q4_K](https://github.com/ggerganov/ggml/blob/a3c0188a4b5d3dec052ff87c9f773baa53631d70/src/ggml-common.h#L239-L249) , but I'm not sure of its purpose. I didn't see it being used in quantization or dequantization. What is the purpose of this value? Is it for memory alignment or something?

**Comment by ggerganov** [comment, marked as answer]
It's used in GPU code that supports `half2` ops. For example in CUDA:

https://github.com/ggerganov/ggml/blob/a3c0188a4b5d3dec052ff87c9f773baa53631d70/src/ggml-cuda/fattn-common.cuh#L120

Since this is a `union`, the `dm` member is basically an alias for the `d` and `dmin` factors

**Reply by PenutChen** [reply]
Thanks for explaining, really cool tips!
